File: solutions/elixir/149-counting-bits.ex
# Problem 149: Counting Bits (LeetCode 338)
# Difficulty: Easy
# Language: Elixir
# 
defmodule CountBits do
  def count_bits(n) do
    Enum.reduce(0..n, %{}, fn i, dp ->
      Map.put(dp, i, if(i == 0, do: 0, else: dp[div(i, 2)] + rem(i, 2)))
    end) |> Enum.sort() |> Enum.map(&elem(&1, 1))
  end
end

File: solutions/typescript/149-counting-bits.ts
/**
 * Problem 149: Counting Bits (LeetCode 338)
 * Difficulty: Easy
 * Language: TypeScript
 */
function countBits(n: number): number[] {
  return Array.from({ length: n + 1 }, (_, i) => i).reduce<number[]>(
    (dp, i) => [...dp, i === 0 ? 0 : dp[i >> 1] + (i & 1)], []);
}
